lib/cluster/subscriber.js

```diff
@@ -0,0 +1,699 @@
+    const { collection, index, scope } = message;
+
+    debug(
+      'New collection added by node %s (scope: %s, index: %s, collection: %s)',
+      this.remoteNodeId,
+      scope,
+      index,
+      collection);
+
+    await global.kuzzle.ask(
+      `core:storage:${scope}:cache:addCollection`,
+      index,
+      collection);
+  }
+
+  /**
+   * Removes a collection from the index cache
+   *
+   * @param {Object} message - decoded IndexCacheAdd protobuf message
+   * @return {void}
+   */
+  async handleCollectionRemoval (message) {
+    const { collection, index, scope } = message;
+
+    debug(
+      'Indexes removed by node %s (scope: %s, index: %s, collection: %s)',
+      this.remoteNodeId,
+      scope,
+      index,
+      collection);
+
+    await global.kuzzle.ask(
+      `core:storage:${scope}:cache:removeCollection`,
+      index,
+      collection);
+  }
+
+
+  /**
+   * Checks that we did receive a heartbeat from the remote node
+   * If a heartbeat is missing, we allow 1 heartbeat round for the remote node
+   * to recover, otherwise we evict it from the cluster.
+   */
+  async checkHeartbeat () {
+    if (this.state === stateEnum.EVICTED) {
+      return;
+    }
+
+    const now = Date.now();
+
+    if ((now - this.lastHeartbeat) > this.heartbeatDelay) {
+      if (this.state === stateEnum.MISSING_HEARTBEAT) {
+        this.state = stateEnum.EVICTED;
+
+        await this.evictNode({
+          broadcast: true,
+          reason: 'heartbeat timeout',
+        });
+      }
+      else {
+        this.state = stateEnum.MISSING_HEARTBEAT;
+      }
+    }
+    else {
+      this.state = stateEnum.SANE;
+    }
+  }
+
+  /**
+   * Disconnects from the remote node, and frees all allocated resources.
+   */
+  dispose () {
+    if (this.state === stateEnum.EVICTED) {
+      return;
+    }
+
+    this.state = stateEnum.EVICTED;
+    this.socket.close();
+    this.socket = null;
+    clearInterval(this.heartbeatTimer);
+  }
+
+  /**
+   * Checks that the received message is the one we expect.
+   * @param  {Object} message - decoded protobuf message
+   * @return {boolean} false: the message must be discarded, true otherwise
+   */
+  async validateMessage (message) {
+    if (!has(message, 'messageId')) {
+      global.kuzzle.log.warn(`Invalid message received from node ${this.remoteNodeId}. Evicting it.`);
+
+      await this.evictNode({
+        broadcast: true,
+        reason: 'invalid message received (missing "messageId" field)',
+      });
+
+      return false;
+    }
+
+    if ( this.state === stateEnum.BUFFERING
+      && this.lastMessageId.greaterThanOrEqual(message.messageId)
+    ) {
+      return false;
+    }
+
+    this.lastMessageId = this.lastMessageId.add(1);
+
+    if (this.lastMessageId.notEquals(message.messageId)) {
+      global.kuzzle.log.error(`[FATAL] Node out-of-sync: ${message.messageId - this.lastMessageId - 1} messages lost from node ${this.remoteNodeId}`);
+      global.kuzzle.shutdown();
```

Comment:
Can't we broadcast a node eviction message to save a round of heartbeat to other nodes?

Reply:
Resolved (see above)